src/chocolatey/infrastructure/registration/AssemblyResolution.cs

```diff
@@ -17,26 +17,95 @@
 {
     using System;
+    using System.Collections.Concurrent;
     using System.Linq;
     using System.Threading;
     using adapters;
+    using chocolatey.infrastructure.app;
     using filesystem;
 
     public class AssemblyResolution
     {
         private const int LOCK_RESOLUTION_TIMEOUT_SECONDS = 5;
         private static readonly object _lockObject = new object();
 
+        private static readonly ConcurrentDictionary<string, IAssembly> _loadedAssemblies = new ConcurrentDictionary<string, IAssembly>();
+
         /// <summary>
         /// Resolves or loads an assembly. If an assembly is already loaded, no need to reload it.
         /// </summary>
         /// <param name="assemblySimpleName">Simple Name of the assembly, such as "chocolatey"</param>
         /// <param name="publicKeyToken">The public key token.</param>
         /// <param name="assemblyFileLocation">The assembly file location. Typically the path to the DLL on disk.</param>
+        /// <param name="ignoreExisting">Whether any existing library that has previously been loaded should be ignored or not.</param>
         /// <returns>An assembly</returns>
         /// <exception cref="Exception">Unable to enter synchronized code to determine assembly loading</exception>
-        public static IAssembly resolve_or_load_assembly(string assemblySimpleName, string publicKeyToken, string assemblyFileLocation)
+        public static IAssembly resolve_or_load_assembly(string assemblySimpleName, string publicKeyToken, string assemblyFileLocation, bool ignoreExisting = false)
+        {
+            var lockTaken = false;
+            try
+            {
+                Monitor.TryEnter(_lockObject, TimeSpan.FromSeconds(LOCK_RESOLUTION_TIMEOUT_SECONDS), ref lockTaken);
+            }
+            catch (Exception)
+            {
+                throw new Exception("Unable to enter synchronized code to determine assembly loading");
+            }
+
+            IAssembly resolvedAssembly = null;
+
+            if (lockTaken)
+            {
+                try
+                {
+                    if (!ignoreExisting)
+                    {
+                        resolvedAssembly = resolve_assembly(assemblySimpleName, publicKeyToken);
+                    }
+
+                    if (resolvedAssembly == null)
+                    {
+                        var tempAssembly = Assembly.Load(FileSystem.read_binary_file_into_byte_array(assemblyFileLocation));
+
+                        if (tempAssembly == null)
+                        {
+                            return null;
+                        }
+
+                        if (string.IsNullOrWhiteSpace(publicKeyToken) || tempAssembly.GetName().get_public_key_token().is_equal_to(publicKeyToken))
+                        {
+                            "AssemblyResolver".Log().Debug("Loading up '{0}' assembly type from '{1}'".format_with(assemblySimpleName, assemblyFileLocation));

```

Comment:
I can change it to use `chocolatey`, but it doesn't matter as the debug log at this stage won't actually show up anyway.

Technically it could be removed, but I left it in so we have it if we track down why debug messaging doesn't work at this point in the future (btw, already tried using `chocolatey` which didn't work either).